service/src/DCGMIOGroup.cpp

```diff
@@ -0,0 +1,541 @@
+        m_dcgm_device_pool.max_storage_time(10); //10 seconds.
+    }
+
+    void DCGMIOGroup::save_control(const std::string &save_path)
+    {
+        throw Exception("DCGMIOGroup::save_control()",
+                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
+    }
+
+    void DCGMIOGroup::restore_control(const std::string &save_path)
+    {
+        throw Exception("DCGMIOGroup::restore_control()",
+                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
+    }
+
+    // Hint to Agent about how to aggregate signals from this IOGroup
+    std::function<double(const std::vector<double> &)> DCGMIOGroup::agg_function(const std::string &signal_name) const
+    {
+        auto it = m_signal_available.find(signal_name);
+        if (it == m_signal_available.end()) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + signal_name +
+                            "not valid for DCGMIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return it->second.m_agg_function;
+    }
+
+    // Specifies how to print signals from this IOGroup
+    std::function<std::string(double)> DCGMIOGroup::format_function(const std::string &signal_name) const
+    {
+        auto it = m_signal_available.find(signal_name);
+        if (it == m_signal_available.end()) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + signal_name +
+                            "not valid for DCGMIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return it->second.m_format_function;
+    }
+
+    // A user-friendly description of each signal
+    std::string DCGMIOGroup::signal_description(const std::string &signal_name) const
+    {
+        if (!is_valid_signal(signal_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": signal_name " + signal_name +
+                            " not valid for DCGMIOGroup.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return m_signal_available.at(signal_name).m_description;
+    }
+
+    // A user-friendly description of each control
+    std::string DCGMIOGroup::control_description(const std::string &control_name) const
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name +
+                            "not valid for DCGMIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        return m_control_available.at(control_name).m_description;
+    }
+
+    int DCGMIOGroup::signal_behavior(const std::string &signal_name) const
+    {
+        return IOGroup::M_SIGNAL_BEHAVIOR_VARIABLE;
+    }
+
+    // Name used for registration with the IOGroup factory
+    std::string DCGMIOGroup::plugin_name(void)
+    {
+        return "dcgm";
```

Comment:
We use all-caps for the plugin_name() for all IOGroups except the ones you wrote @lhlawson.  Sorry I'm just catching this now, but for consistency, we should return an all-caps plugin_name for DGGMIOGroup, LevelZeroIOGroup and the NVMLIOgroup.

Reply:
Addressed for DCGM as part of this PR.  I'll put up a separate PR to update other IOGroups to LEVELZERO and NVML.

Reply:
#2065 filed to track NVML and LEVELZERO change